High memory usage

#### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
PyCharm 2023.1.4 (Community Edition)
Build #PC-231.9225.15, built on July 11, 2023
Runtime version: 17.0.7+10-b829.16 aarch64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.

Non-Bundled Plugins:
    com.chesterccw.excelreader (2023.6.1-222.232)
    org.rust.lang (0.4.198.5409-231)
    com.intellij.plugin.adernov.powershell (2.0.10)
    nl.rubensten.texifyidea (0.7.32)
    com.firsttimeinforever.intellij.pdf.viewer.intellij-pdf-viewer (0.15.0)


#### Operating System 
macOS 13.4.1

GC: G1 Young Generation, G1 Old Generation
Memory: 3072M (Currently allocated for PyCharm, it was up to 6GB out of 16GB but still crashed)
Cores: 8
Metal Rendering is ON
Registry:
    debugger.new.tool.window.layout=true
    ide.experimental.ui=true

#### TeXiFy IDEA version
0.7.32

#### What I did (steps to reproduce)
Opening a new project, open the main.tex and wait, no matter how much memory I give to PyCharm while a LaTex document is open it will run out of memory and the IDE will become unresponsive and eventually have to be killed by Activity Monitor if it doesn't crash on it's own.

#### Minimal example to reproduce the problem
<!-- Make sure you have read and understood how to create a minimal example: https://tex.meta.stackexchange.com/questions/228/ive-just-been-asked-to-write-a-minimal-example-what-is-that -->

```latex
\documentclass{article}
\begin{document}
    
\end{document}
```

#### Expected behavior
Not have the IDE run out of memory

#### Actual behavior
The IDE runs out of memory, I started with the default 2GB and increased by 1GB after each incident up to 6GB and it continues to run out only while in a Latex document. Python and Rust projects have no issues. 

#### (if applicable) The full stacktrace of the exception thrown
```

```

<!-- If you see a minor issue related to commands or environments metadata, e.g. which commands are cite or verbatim commands, you are encouraged to try to add the missing information yourself, see https://github.com/Hannah-Sten/TeXiFy-IDEA/wiki/Contributing-to-TeXiFy#editing-magic -->

I installed from disk selected the file (instrumented-TeXiFy-IDEA-0.7.34-alpha.1.jar). Now the IDE keeps getting IDE error occurred. Exception in the pluginin.

java.lang.NoClassDefFoundError: Could not initialize class nl.hannahsten.texifyidea.util.magic.CommandMagic
	at nl.hannahsten.texifyidea.psi.LatexLanguageInjector.getLanguagesToInject(LatexLanguageInjector.kt:63)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageManagerImpl$PsiManagerRegisteredInjectorsAdapter.getLanguagesToInject(InjectedLanguageManagerImpl.java:529)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageManagerImpl.processInPlaceInjectorsFor(InjectedLanguageManagerImpl.java:495)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.probeElementsUpInner(InjectedLanguageUtilBase.java:238)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.lambda$probeElementsUp$0(InjectedLanguageUtilBase.java:218)
	at com.intellij.openapi.application.impl.ReadActionCacheImpl$allowInWriteAction$1.invoke(ReadActionCacheImpl.kt:42)
	at com.intellij.openapi.application.impl.ReadActionCacheImpl$allowInWriteAction$1.invoke(ReadActionCacheImpl.kt:42)
	at com.intellij.openapi.application.impl.ReadActionCacheImpl.allowInWriteAction(ReadActionCacheImpl.kt:30)
	at com.intellij.openapi.application.impl.ReadActionCacheImpl.allowInWriteAction(ReadActionCacheImpl.kt:42)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.probeElementsUp(InjectedLanguageUtilBase.java:217)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.enumerate(InjectedLanguageUtilBase.java:159)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.findInside(InjectedLanguageUtilBase.java:401)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.tryOffset(InjectedLanguageUtilBase.java:380)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtilBase.findInjectedElementNoCommit(InjectedLanguageUtilBase.java:360)
	at com.intellij.psi.impl.source.tree.injected.InjectedLanguageManagerImpl.findInjectedElementAt(InjectedLanguageManagerImpl.java:351)
	at com.intellij.openapi.editor.EditorMouseHoverPopupManager.findElementForQuickDoc(EditorMouseHoverPopupManager.java:354)
	at com.intellij.openapi.editor.EditorMouseHoverPopupManager.findElementForQuickDoc(EditorMouseHoverPopupManager.java:346)
	at com.intellij.openapi.editor.EditorMouseHoverPopupManager.createContext(EditorMouseHoverPopupManager.java:332)
	at com.intellij.openapi.editor.EditorMouseHoverPopupManager.lambda$showInfoTooltip$7(EditorMouseHoverPopupManager.java:405)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$OTelMonitor.callWrapped(NonBlockingReadActionImpl.java:832)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$OTelMonitor$MonitoredComputation.call(NonBlockingReadActionImpl.java:864)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.insideReadAction(NonBlockingReadActionImpl.java:602)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.lambda$attemptComputation$4(NonBlockingReadActionImpl.java:565)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1133)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runInReadActionWithWriteActionPriority$0(ProgressIndicatorUtils.java:76)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtilService.runActionAndCancelBeforeWrite(ProgressIndicatorUtilService.java:63)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:133)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runWithWriteActionPriority$1(ProgressIndicatorUtils.java:114)
	at com.intellij.openapi.progress.ProgressManager.lambda$runProcess$0(ProgressManager.java:71)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:61)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.ProgressManager.runProcess(ProgressManager.java:71)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runWithWriteActionPriority(ProgressIndicatorUtils.java:111)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runInReadActionWithWriteActionPriority(ProgressIndicatorUtils.java:76)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.attemptComputation(NonBlockingReadActionImpl.java:565)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.lambda$transferToBgThread$1(NonBlockingReadActionImpl.java:466)
	at com.intellij.openapi.application.impl.NonBlockingReadActionImpl$Submission.lambda$transferToBgThread$2(NonBlockingReadActionImpl.java:481)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.NoClassDefFoundError: arrow/core/MapKt [in thread "DefaultDispatcher-worker-9"]
	at nl.hannahsten.texifyidea.lang.commands.LatexRegularCommand.<clinit>(LatexRegularCommand.kt:37)
	at nl.hannahsten.texifyidea.util.magic.CommandMagic.<clinit>(CommandMagic.kt:95)
	at nl.hannahsten.texifyidea.index.file.LatexExternalEnvironmentDataIndexer.<init>(LatexExternalEnvironmentDataIndexer.kt:23)
	at nl.hannahsten.texifyidea.index.file.LatexExternalEnvironmentIndex.<init>(LatexExternalEnvironmentIndex.kt:20)
	at com.intellij.serviceContainer.ComponentManagerImpl.doInstantiateClass(ComponentManagerImpl.kt:962)
	at com.intellij.serviceContainer.ComponentManagerImpl.instantiateClass(ComponentManagerImpl.kt:954)
	at com.intellij.openapi.extensions.impl.XmlExtensionAdapter.instantiateClass(XmlExtensionAdapter.kt:78)
	at com.intellij.openapi.extensions.impl.XmlExtensionAdapter$SimpleConstructorInjectionAdapter.instantiateClass(XmlExtensionAdapter.kt:89)
	at com.intellij.openapi.extensions.impl.XmlExtensionAdapter.doCreateInstance(XmlExtensionAdapter.kt:49)
	at com.intellij.openapi.extensions.impl.XmlExtensionAdapter.createInstance(XmlExtensionAdapter.kt:30)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.processAdapter(ExtensionPointImpl.java:446)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.access$200(ExtensionPointImpl.java:33)
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl$1.next(ExtensionPointImpl.java:354)
	at com.intellij.util.indexing.FileBasedIndexDataInitialization.initAssociatedDataForExtensions(FileBasedIndexDataInitialization.java:75)
	at com.intellij.util.indexing.FileBasedIndexDataInitialization.prepareTasks(FileBasedIndexDataInitialization.java:131)
	at com.intellij.util.indexing.IndexDataInitializer.call(IndexDataInitializer.kt:21)
	at com.intellij.util.indexing.IndexDataInitializer$Companion$submitGenesisTask$1.invokeSuspend(IndexDataInitializer.kt:50)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,4 +1,4 @@
-pluginVersion = 0.7.33
+pluginVersion = 0.7.34-alpha.1
 
 #     Info about build ranges: https://www.jetbrains.org/intellij/sdk/docs/basics/getting_started/build_number_ranges.html
 #    Note that an xyz branch corresponds to version 20xy.z and a since build of xyz.*

--- a/src/nl/hannahsten/texifyidea/util/files/ImportPackage.kt
+++ b/src/nl/hannahsten/texifyidea/util/files/ImportPackage.kt
@@ -12,6 +12,7 @@ import nl.hannahsten.texifyidea.reference.InputFileReference
 import nl.hannahsten.texifyidea.util.appendExtension
 import nl.hannahsten.texifyidea.util.magic.CommandMagic
 import nl.hannahsten.texifyidea.util.parser.requiredParameter
+import kotlin.math.min
 
 /**
  * This method will try to find a file when the 'import' package is used, which means that including files have to be searched for import paths.
@@ -84,7 +85,7 @@ fun checkForAbsolutePath(command: LatexCommands): VirtualFile? {
 }
 
 fun findRelativeSearchPathsForImportCommands(command: LatexCommands, givenRelativeSearchPaths: List<String> = listOf("")): List<VirtualFile> {
-    var relativeSearchPaths = givenRelativeSearchPaths.toMutableList()
+    var relativeSearchPaths = givenRelativeSearchPaths.toMutableSet()
     val allIncludeCommands = LatexIncludesIndex.getItems(command.project)
     // Commands which may include the current file (this is an overestimation, better would be to check for RequiredFileArguments)
     var includingCommands = allIncludeCommands.filter {
@@ -94,12 +95,15 @@ fun findRelativeSearchPathsForImportCommands(command: LatexCommands, givenRelati
         // Only consider commands that can include LaTeX files
         LatexCommand.lookup(includeCommand.name)?.firstOrNull()?.getArgumentsOf(RequiredFileArgument::class.java)?.any { it.supportedExtensions.contains("tex") } == true
     }
+    // To avoid infinite loops, keep track of where we have been
+    val visitedIncludeCommands = includingCommands.toMutableSet()
+    val handledFiles = mutableListOf<PsiFile>()
 
     // Assume the containingFile might be a root file - if there are no includingCommands we will not search further anyway
     val defaultParentDir = command.containingFile.containingDirectory?.virtualFile?.findFileByRelativePath(command.requiredParameter(0) ?: "")
 
-    // Avoid endless loop (in case of a file inclusion loop)
-    val maxDepth = allIncludeCommands.size
+    // Avoid endless loop (in case of a file inclusion loop), limited to some reasonable maximum nesting level
+    val maxDepth = min(allIncludeCommands.size, 20)
     var counter = 0
 
     // Final paths
@@ -110,9 +114,10 @@ fun findRelativeSearchPathsForImportCommands(command: LatexCommands, givenRelati
     }
 
     // I think it's a kind of reversed BFS
+    // Loop over all commands which possibly include the current element
     while (includingCommands.isNotEmpty() && counter < maxDepth) {
-        val newSearchPaths = mutableListOf<String>()
-        val handledFiles = mutableListOf<PsiFile>()
+        counter++
+        val newSearchPaths = mutableSetOf<String>()
         val newIncludingcommands = mutableListOf<LatexCommands>()
 
         for (includingCommand in includingCommands) {
@@ -133,6 +138,7 @@ fun findRelativeSearchPathsForImportCommands(command: LatexCommands, givenRelati
             // Find files/commands to search next
             val file = includingCommand.containingFile
             if (file !in handledFiles) {
+                handledFiles.add(file)
                 val commandsIncludingThisFile = allIncludeCommands.filter { includeCommand -> includeCommand.getRequiredParameters().any { it.contains(file.name) } }
                 if (commandsIncludingThisFile.isEmpty()) {
                     // Cool, we supposedly have a root file, now try to find the directory containing the file being included by command
@@ -142,13 +148,13 @@ fun findRelativeSearchPathsForImportCommands(command: LatexCommands, givenRelati
                     }
                 }
                 newIncludingcommands.addAll(commandsIncludingThisFile)
-                handledFiles.add(file)
             }
         }
 
-        includingCommands = newIncludingcommands
+        val unvisitedNewIncludingCommands = newIncludingcommands.filter { !visitedIncludeCommands.contains(it) }
+        includingCommands = unvisitedNewIncludingCommands
+        visitedIncludeCommands.addAll(unvisitedNewIncludingCommands)
         relativeSearchPaths = newSearchPaths
-        counter++
     }
 
     return absoluteSearchDirs

